File: credit.c
#include <cs50.h>
#include <stdio.h>

int luhns(long number);
void kind_of_card(long number);

int main(void)
{
    long card_number = get_long("Number: ");
    int sum = luhns(card_number);
    sum %= 10;
    if (sum == 0)
    {
        kind_of_card(card_number);
    }
    else
    {
        printf("INVALID\n");
    }
}

int luhns(long number)
{
    int counter = 0, sum = 0, sum2 = 0;
    while (number > 0)
    {
        // Para separar o ultimo digito
        int last_digit = number % 10;
        // Para excluir o ultimo digito
        number /= 10;
        if (counter == 1)
        {
            int doubling;
            doubling = last_digit * 2;
            if (doubling > 9)
            {
                sum = sum + 1 + (doubling % 10);
            }
            else
            {
                sum = sum + doubling;
            }
            counter = 0;
        }
        else
        {
            sum2 = sum2 + last_digit;
            counter++;
        }
    }
    sum += sum2;
    return sum;
}

void kind_of_card(long number)
{
    int counter = 0;
    long num;
    num = number;
    while (num > 0)
    {
        int last_digit = num % 10;
        num /= 10;
        counter++;
    }
    int first_digit;
    num = number;
    while (num >= 10)
    {
        num /= 10;
    }
    first_digit = num;
    num = number;
    int first_two_digits;
    while (num >= 100)
    {
        num /= 10;
    }

    first_two_digits = num;

    if ((first_digit == 4 && counter == 13) || (first_digit == 4 && counter == 16))
    {
        printf("VISA\n");
    }
    else if ((first_two_digits == 34 && counter == 15) || (first_two_digits == 37 && counter == 15))
    {
        printf("AMEX\n");
    }
    else if (first_two_digits >= 51 && first_two_digits <= 55 && counter == 16)
    {
        printf("MASTERCARD\n");
    }
    else
    {
        printf("INVALID\n");
    }
}

File: credit.py
from cs50 import get_int
import re

def main():
    card_number = get_int("Number: ");
    print(kind_of_card(card_number))


def kind_of_card(n):
# Regex to determine credit card type

    visa_pattern = r"^4\d{12}(\d{3})?(\d{3})?$"  # VISA cards start with 4 and can be 13 or 16 digits
    mastercard_pattern = r"^5[1-5]\d{14}$"  # MASTERCARD starts with 51-55 and is 16 digits
    amex_pattern = r"^3[47]\d{13}$"  # AMEX cards start with 34 or 37 and are 15 digits

    # Convert int to string because 're.match' expects string
    n = str(n)
    
    # Match card number and patterns
    if re.match(visa_pattern, n):
        return "VISA"
    elif re.match(mastercard_pattern, n):
        return "MASTERCARD"
    elif re.match(amex_pattern, n):
        return "AMEX"
    else:
        return "INVALID"


if __name__ == "__main__":
    main()
